content/blog/unit-test-infrastructure/index.md

```diff
@@ -0,0 +1,314 @@
+    }
+}
+
+func (mocks) Call(token string, args resource.PropertyMap, provider string) (resource.PropertyMap, error) {
+    return args, fmt.Errorf("unexpected function call %v", token)
+}
+
+// Applying unit tests.
+func TestInfrastructure(t *testing.T) {
+    err := pulumi.RunErr(func(ctx *pulumi.Context) error {
+        infra, err := createInfrastructure(ctx)
+        assert.NoError(t, err)
+
+        var wg sync.WaitGroup
+        wg.Add(4)
+
+             // Test if the service has tags and a name tag.
+        pulumi.All(infra.server.URN(), infra.server.Tags).ApplyT(func(all []interface{}) error {
+            urn := all[0].(pulumi.URN)
+            tags := all[1].(map[string]interface{})
+
+            assert.Containsf(t, tags, "name", "missing a name tag on server %v", urn)
+            wg.Done()
+            return nil
+        })
+
+             // Test if the instance is configured with user_data.
+        pulumi.All(infra.server.URN(), infra.server.UserData).ApplyT(func(all []interface{}) error {
+            urn := all[0].(pulumi.URN)
+            userData := all[1].(*string)
+
+            assert.Nilf(t, userData, "illegal use of userData on server %v", urn)
+            wg.Done()
+            return nil
+        })
+
+             // Test if service is a member of a security group.
+        pulumi.All(infra.server.URN(), infra.server.SecurityGroups).ApplyT(func(all []interface{}) error {
+            urn := all[0].(pulumi.URN)
+            securityGroups := all[1].([]string)
+
+            assert.Greaterf(t, len(securityGroups), 0, "illegal security group spec on server %v", urn)
+            wg.Done()
+            return nil
+        })
+
+             // Test if port 22 for ssh is exposed.
+        pulumi.All(infra.group.URN(), infra.group.Ingress).ApplyT(func(all []interface{}) error {
+            urn := all[0].(pulumi.URN)
+            ingress := all[1].([]ec2.SecurityGroupIngress)
+
+            for _, i := range ingress {
+                openToInternet := false
+                for _, b := range i.CidrBlocks {
+                    if b == "0.0.0.0/0" {
+                        openToInternet = true
+                        break
+                    }
+                }
+
+                assert.Falsef(t, i.FromPort == 22 && openToInternet, "illegal SSH port 22 open to the Internet (CIDR 0.0.0.0/0) on group %v", urn)
+            }
+
+            wg.Done()
+            return nil
+        })
+
+        wg.Wait()
+        return nil
+    }, pulumi.WithMocks("project", "stack", mocks(0)))
+    assert.NoError(t, err)
+}
+```
+
+## Try it out today
+
+We’ve released these features early so that you can try them out before the 2.0 release. With the 2.0 release we’ll have more documentation, examples, and blog posts to get you started with writing and running tests for your infrastructure. In the meantime, we would appreciate feedback on our [Community Slack channel](https://pulumi-community.slack.com/).
```

Comment:
I merged the examples: https://github.com/pulumi/examples#testing

Reply:
Thanks! I'll add a link to them